### PR TITLE
Add translation update scripts to integrate with localization platforms

### PR DESCRIPTION
**Pull request by guoyunhe**
Many new strings were added to `intl/msg_hash_us.h` but they are not updated to other translations. Manually checking these files is impossible. So I created this script to insert new messages and remove out-dated messages from translation files.

Here is one blocker to integrate translation systems: the message key is not unique because here are some `#if` / `#else` conditionally define the same key with different messages. I think the condition shouldn't be in message files but in where they are used. So I renamed duplicated keys in this PR.

The workflow is:

## Steps

- Developers update strings in `msg_hash_us.h`.
- Developers (can set a cron job) run `./h2json.py msg_hash_us.h` to generate `msg_hash_us.json`. It is just a convenient format that is supported by Weblate/Crowdin/Transifex and doesn't need to be inversion control.
- Developers (can set a cron job) upload `msg_hash_us.json` to Weblate/Crowdin/Transifex.
- Translators translate strings on Weblate/Crowdin/Transifex.
- Developers (can set a cron job) download `msg_hash_xx.json` files.
- Developers (can set a cron job) run `./json2h.py msg_hash_xx.json` to generate `msg_hash_xx.h`.

## Pros

- No new dependencies.
- No performance impact.
- Don't require translators to know how to use Git, how to read C code and how to create Pull Request.
- Translators will be informed whenever a source string changes.

## Example integration with Crowdin

https://crowdin.com/project/retroarch

### CODE REVIEW

File: config.def.keybinds.h [hunk 1]
```diff
@@ -153,7 +157,11 @@ static const struct retro_keybind retro_keybinds_1[] = {
    { true, RARCH_LOAD_STATE_KEY,           MENU_ENUM_LABEL_VALUE_INPUT_META_LOAD_STATE_KEY,        RETROK_F4,        NO_BTN, NO_BTN, 0, AXIS_NONE, AXIS_NONE, AXIS_NONE, NULL, NULL },
    { true, RARCH_SAVE_STATE_KEY,           MENU_ENUM_LABEL_VALUE_INPUT_META_SAVE_STATE_KEY,        RETROK_F2,        NO_BTN, NO_BTN, 0, AXIS_NONE, AXIS_NONE, AXIS_NONE, NULL, NULL },
    { true, RARCH_FULLSCREEN_TOGGLE_KEY,    MENU_ENUM_LABEL_VALUE_INPUT_META_FULLSCREEN_TOGGLE_KEY, RETROK_f,         NO_BTN, NO_BTN, 0, AXIS_NONE, AXIS_NONE, AXIS_NONE, NULL, NULL },
+#ifdef HAVE_LAKKA
+   { true, RARCH_QUIT_KEY,                 MENU_ENUM_LABEL_VALUE_INPUT_META_RESTART_KEY,           RETROK_ESCAPE,    NO_BTN, NO_BTN, 0, AXIS_NONE, AXIS_NONE, AXIS_NONE, NULL, NULL },
```
Comment (RobLoach) [inline config.def.keybinds.h]: This question is out of scope, but should we introduce a `RARCH_RESTART_KEY` to save having needing the `HAVE_LAKKA` conditional?

Comment (guoyunhe) [reply config.def.keybinds.h]: Will be good to have.

**Comment by guoyunhe** [comment]
The sample integration with Crowdin is working well. Welcome to try it out and give some opinions!

https://crowdin.com/project/retroarch

@theheroGAC @IlDucci @fpscan @WeedyWeedSmoker @DEX357

**Comment by IlDucci** [comment]
Can you please update the non-English languages to their current version?

**Comment by guoyunhe** [comment]
@IlDucci they should be up-to-date. Some strings that have been removed from English were also removed in non-English version.

**Comment by twinaphex** [comment]
OK, this is a pretty great solution. If we merge this, then are we (as Libretro/RetroArch mods) already setup to be able to use this on crowdin.com? (with regards to this step - Developers (can set a cron job) upload msg_hash_us.json to Weblate/Crowdin/Transifex.)

**Comment by guoyunhe** [comment]
The cron job setup process is easy:

1. [Install crowdin cli tool](https://support.crowdin.com/cli-tool/)
2. Get the `intl/crowdin.yaml` file (I can send you by email)
3. Add cron job `cd path/to/retroarch/intl && ./crowin_sync.sh`

**Comment by IlDucci** [comment]
> Crowdin is really cool, I used to help some projects in there but main issue is; people translates sentences without checking origin(like where it will be use, what is background etc.).

That happens when some random guy goes in and translate. Crowdin is a valuable tool, for sure, but it needs checks and balances.

Another thing that might be needed is the creation of a "glossary" with the most used terms, or those that were discussed by previous translators (might want to scrape the Issues/Pull Requests of this Git).

**Comment by guoyunhe** [comment]
Ideally, each language should have a reviewer. Only approved translations go into code base. Crowdin makes it easier to proof read strings. Admins can one click undo all random changes by a random guy.

**Comment by IlDucci** [comment]
Another thing I would like to ask, regarding the lack of origin: Is it possible to link the entry names with the descriptions automatically or is that a manual process?

**Comment by guoyunhe** [comment]
Maybe this screenshot can explain it:

image

You will be able to see both key and source string of an entry in the editor.

**Comment by gouchi** [comment]
> Crowdin is really cool, I used to help some projects in there but main issue is; people translates sentences without checking origin(like where it will be use, what is background etc.).

I like how they have done it with [Telegram](https://translations.telegram.org/en/) where you have the interface and translation side by side.

Anyway using Crowdin is nice improvement for the translation !

Thank you @guoyunhe

**Comment by IlDucci** [comment]
I have a couple of questions:
 - Is this going to be the definitive repo, or should I wait until the team picks up the admin rights or makes a new one?
 - Are you planning on covering the .c file, the one with the help screens? If so, do you know if Crowdin supports character and line limits?

**Comment by guoyunhe** [comment]
1. You can wait until this PR is merged.
2. It is impossible to cover .c files because they are not key-value pair format.

**Comment by IlDucci** [comment]
>     1. You can wait until this PR is merged.

I'm okay with that...

>     2. It is impossible to cover .c files because they are not key-value pair format.

But not with this. If you are asking to switch platforms, wouldn't it make more sense to do it so every localizable string is on the new one?

**Comment by guoyunhe** [comment]
.c files can be rewritten to a more friendly format and it can a stage 2 target. only if stage 1 target is done, i can continue improve it. it won't be a problem in future.

**Comment by guoyunhe** [comment]
@twinaphex do you think this solution is ready to go? if here are anything need to be improved/solved, let me know!

**Comment by nextlooper42** [comment]
Can i start with translation on https://crowdin.com/project/retroarch or should i wait?

**Comment by guoyunhe** [comment]
@nextlooper42 please wait. I cannot guarantee that the PR can be merged. Still waiting for developers' approvals.

**Comment by twinaphex** [comment]
@guoyunhe 

OK, if you can make me a manager here -

https://crowdin.com/project/retroarch

We can go along with this and start movign forward with this.

**Comment by guoyunhe** [comment]
@twinaphex what is your username?

**Comment by guoyunhe** [comment]
@twinaphex done.

**Comment by WeedyWeedSmoker** [comment]
@twinaphex I just tried it, but there's a MENU_ENUM_LABEL_VALUE_INPUT_META_RESTART_KEY that is translated neither by me, nor the spanish translator(s) that I can't find anywhere, not even in the US file…

Do you happen to know where does it come from?

**Comment by twinaphex** [comment]
OK, merged now. 

@guoyunhe Let me know of any followup steps that I still have to take.

**Comment by guoyunhe** [comment]
@twinaphex you can test the cron job (maybe in a test branch). then check https://crowdin.com/project/retroarch/activity_stream (activities are updated 5-10 minutes later).

@WeedyWeedSmoker that is a new key I introduced in this PR to avoid some #if #else issues.

**Comment by IlDucci** [comment]
I've updated the Spanish sheet in Crowdin and started to add some comments related on certain terms.

I wanted to get notifications for this Crowdin project whenever there's new strings, but it seems like enabling that feature is restricted to the project creator. Could you please enable this function?

**Comment by guoyunhe** [comment]
@IlDucci enabled
image

**Comment by IlDucci** [comment]
Thank you.

**Comment by twinaphex** [comment]
@guoyunhe Can't get it to work-

D:\Apps\Crowdin>setup_crowdin.bat
ERROR: Access to the registry path is denied.

WARNING: The data being saved is truncated to 1024 characters.
ERROR: Access to the registry path is denied.
FIND: 'java version': No such file or directory
1 was unexpected at this time.

I have Java installed.

**Comment by twinaphex** [comment]
Also, I mainly use mingw. Is this going to work with mingw?

I have to say it really sucks that their CLI tool is Java-only.

**Comment by guoyunhe** [comment]
Haven't try Windows version. I am using Linux.

@twinaphex Which installer did you download? Is it the exe here? https://downloads.crowdin.com/cli/v2/crowdin.exe

**Comment by twinaphex** [comment]
Yeah I installed that.

Anyway, I'm still unable to use this. Please provide clear instructions, step by step, on how to use this on Windows. It keeps complaining that Java is not installed even though I install it manually over and over again.

**Comment by guoyunhe** [comment]
I am installing Windows and will keep you updated.

**Comment by guoyunhe** [comment]
Hi @twinaphex , here is the steps to run crowdin. You installed java correctly but the command is missing some parameters. If you open cmd.exe and run:

```bat
cd path\to\retroarch\intl

# Generate *.json file
python3 h2json.py msg_hash_us.h

# Upload sources
java -jar path\to\crowdin-cli.jar upload sources

# Wait for a minute for the server to process data

# Download translations
java -jar path\to\crowdin-cli.jar download translations

# Generate *.h files
python3 json2h.py msg_hash_fr.json
python3 json2h.py msg_hash_chs.json
python3 json2h.py msg_hash_vn.json
...
```

**Comment by guoyunhe** [comment]
Hi @twinaphex , is the translation script work?

**Comment by twinaphex** [comment]
Hi there, let me get back to you on this within the next few days.

**Comment by fpscan** [comment]
Hi @guoyunhe I haven't setup my pc and network connection isn't available right now. I haven't checked this but I really want to try this but I wondering how this effect the `core_option`.int, so each core will have their own crowdin? Did you had any chance to talk this with Autechre?

**Comment by guoyunhe** [comment]
@fpscan the crowdin project support multiple files. I can add another script for core_option. And contributors can translate core_option files in the same project.

**Comment by fpscan** [comment]
@guoyunhe Hi,

I have some time now, so I am checking Crowding Turkish translations. Should I just upvote the strings you added or do I need to something else?

**Comment by guoyunhe** [comment]
@fpscan I add you as proof reader. now you can approve translations directly.

**Comment by IlDucci** [comment]
Is there a timeline or issues that need to be solved in order to push Crowdin's translations? I just realized there's an alternate Spanish translation from someone who went through the direct, PR route.

**Comment by guoyunhe** [comment]
@IlDucci ideally all translation should be done on Crowdin. If someone still use PRs, please inform them to join Crowdin!

**Comment by IlDucci** [comment]
That doesn't help much if the work that is being done in Crowdin isn't being exported back to RetroArch's repo.

**Comment by guoyunhe** [comment]
It will. You can watch the #10358

**Comment by fpscan** [comment]
@guoyunhe Hi,

Can you add me as Manager or something, @twinaphex knows me. I am going to make some post and video later on this month, so people will know that we have moved Crowdin instead github repo.

**Pull request merged**